### PR TITLE
Add hasattr for tensor variable

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131008
* #131007

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -333,6 +333,23 @@ def method_attr__version(self, tx):
             tx, [self], {}
         )
 
+    def call_hasattr(self, tx, name):
+        try:
+            self.var_getattr(tx, name)
+            ret_val = True
+        except (NotImplementedError, AttributeError):
```
Comment (anijain2305) [inline torch/_dynamo/variables/tensor.py:340]: Are you sure about this? `var_getattr` is Dynamo codebase, are we supposed to return AttributeError. If there is an overridden `__getattr__` function, we might return `ObservedExpection`

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -333,6 +333,23 @@ def method_attr__version(self, tx):
...
+            ret_val = False
+
+        if self.source:
+            install_guard(
+                AttrSource(self.source, name).make_guard(GuardBuilder.HASATTR)
+            )
+            ret_val = ret_val or name in self.proxy.node.meta.get("tensor_dict", dict())
```
Comment (anijain2305) [inline torch/_dynamo/variables/tensor.py:347]: Maybe this should go in `var_getattr`

```
self.proxy.node.meta.get("tensor_dict", dict())
```

File: torch/_dynamo/variables/tensor.py [hunk 3]
```diff
@@ -333,6 +333,23 @@ def method_attr__version(self, tx):
...
+
+        return ConstantVariable(
+            ret_val or tx.output.side_effects.has_pending_mutation_of_attr(self, name)
```
Comment (anijain2305) [inline torch/_dynamo/variables/tensor.py:350]: Wouldn't `tx.output.side_effects.has_pending_mutation_of_attr(self, name)` be already checked by `var_getattr`? If not, should we move it there.

File: torch/_dynamo/variables/tensor.py (hunk 2 above)
Comment (mlazos) [reply torch/_dynamo/variables/tensor.py:347]: oh sure I can add this in there

File: torch/_dynamo/variables/tensor.py (hunk 3 above)
Comment (mlazos) [reply torch/_dynamo/variables/tensor.py:350]: Yeah you're right I think I was just being careful, let me see.

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (mlazos) [reply torch/_dynamo/variables/tensor.py:340]: https://github.com/pytorch/pytorch/blob/82242a258a7728289caf64dfaabef552a2c994b8/torch/_dynamo/variables/tensor.py#L260

Comment (mlazos) [reply torch/_dynamo/variables/tensor.py:340]: If the getattr is defined we return `NotImplementedError`

File: torch/_dynamo/variables/tensor.py (hunk 2 above)
Comment (mlazos) [reply torch/_dynamo/variables/tensor.py:347]: Actually I'm going to remove, this is equivalent to evaluating the input source and checking its attributes which we already do.

**Review by anijain2305 — APPROVED** [review]

**Comment by mlazos** [comment]
@pytochbot merge

**Pull request closed without merging**